File: ros2-for-beginners/ros2_ws/src/my_py_pkg/my_py_pkg/robot_news_station.py
#!/usr/bin/python3
import rclpy
from rclpy.node import Node

# import the messge type from example interfaces
from example_interfaces.msg import String


class RobotNewsStationNode(Node):
    def __init__(self) -> None:
        super().__init__("robot_news_station")

        # parameters
        self.declare_parameter("robot_name", "robot")
        self.declare_parameter("publish_frequency", 1.0)

        self.robot_name_ = self.get_parameter("robot_name").value
        self.publish_frequency_ = self.get_parameter(
            'publish_frequency').value

        '''
        create a publisher object
        using create_publisher method of Node class 
        takes in 3 arguements 
            1. message type.
            2. topic to which it publishes.
            3. queue size -- to keep some message before discarding the previous ones, acts as buffer,
                it allows the application to keep some messages before they are processed.
        '''
        self.publisher_ = self.create_publisher(String, "robot_news", 10)

        # create a timer to publish the message in regular intervals
        self.timer_ = self.create_timer(
            1.0/self.publish_frequency_, self.publish_news)

        # log
        self.get_logger().info("Robot News Station has been started.")

    # create a function to publish the news
    def publish_news(self):
        msg = String()
        msg.data = 'Hello from ' + self.robot_name_ + ", from robot news"
        self.publisher_.publish(msg)


def main(args=None):
    rclpy.init()
    node = RobotNewsStationNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros2-for-beginners/ros2_ws/src/my_cpp_pkg/src/robot_news_station.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"

class RobotNewsStationNode: public rclcpp::Node{
    public:
        RobotNewsStationNode() : Node("robot_news_station") ,robot_name_("R2D2"){
            /*
            initialize publisher
            takes in two parameters 
                1. topic name
                2. queue size

                the message type is already specified in the create_publisher template
            */ 
            
            publisher_ = this->create_publisher<example_interfaces::msg::String>("robot_news",10);

            // initialize timer 
            timer_ = this->create_wall_timer(std::chrono::milliseconds(500),
            std::bind(&RobotNewsStationNode::publishNews,this));

            // node start log 
            RCLCPP_INFO(this->get_logger(),"cpp robot news station has been started");
        };

    private:
    // declarations
    rclcpp::Publisher<example_interfaces::msg::String>::SharedPtr publisher_; //publisher 
    std::string robot_name_; //robot name
    rclcpp::TimerBase::SharedPtr timer_; //timer

    // function to publish
    void publishNews(){
        auto msg = example_interfaces::msg::String();
        msg.data = std::string("Hi this is "+robot_name_+"from robot news");
        publisher_->publish(msg);
    };


};


int main(int argc, char **argv) {
    rclcpp::init(argc,argv);
    auto node = std::make_shared<RobotNewsStationNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}